File: examples/moe.py
#!/usr/bin/env python3
"""
Mixture of Experts (MoE) example - cuTile Python

Fused MoE kernel that multiplies routed tokens by expert weights,
plus a SiLU-and-mul activation kernel. Based on cuTile Python's MoE.py sample.
"""

import cupy as cp
import numpy as np
import cuda.tile as ct
from math import ceil

ConstInt = ct.Constant[int]
ConstBool = ct.Constant[bool]


#=============================================================================
# Helper: 2D swizzle
#=============================================================================

def swizzle_2d(M, N, tm, tn, GROUP_SIZE_M):
    """Get the global IDs of the current block in a 1D grid."""
    bid = ct.bid(0)
    num_bid_m = ct.cdiv(M, tm)
    num_bid_n = ct.cdiv(N, tn)
    num_bid_in_group = GROUP_SIZE_M * num_bid_n
    group_id = bid // num_bid_in_group
    first_bid_m = group_id * GROUP_SIZE_M
    group_size_m = min(num_bid_m - first_bid_m, GROUP_SIZE_M)
    bid_m = first_bid_m + (bid % group_size_m)
    bid_n = (bid % num_bid_in_group) // group_size_m
    return bid_m, bid_n


#=============================================================================
# Kernels
#=============================================================================

@ct.kernel
def fused_moe_kernel(
    A, B, C,
    topk_weights,
    sorted_token_ids,
    sorted_expert_ids,
    num_token_replicas: int,
    mul_routed_weight: ConstBool,
    TILE_M: ConstInt,
    TILE_N: ConstInt,
    TILE_K: ConstInt,
):
    """
    Fused MoE kernel: multiplies routed tokens by their assigned expert weights.

    Token ids are sorted and padded so each expert processes a multiple of TILE_M tokens.
    """
    M = sorted_token_ids.shape[0]
    N = B.shape[1]
    K = B.shape[2]

    GROUP_SIZE_M = 8
    bid_m, bid_n = swizzle_2d(M, N, TILE_M, TILE_N, GROUP_SIZE_M)

    zero_pad = ct.PaddingMode.ZERO

    # Gather replicated/padded token indices for this block
    token_id_indices = bid_m * TILE_M + ct.arange(TILE_M, dtype=ct.int32)
    token_ids = ct.gather(sorted_token_ids, token_id_indices)

    # Collapse replica dimension to recover source row in A
    a_row_indices = token_ids // num_token_replicas

    # Each TILE_M block is homogeneous in expert assignment
    expert_id = ct.load(sorted_expert_ids, index=bid_m, shape=())

    accumulator = ct.full((TILE_M, TILE_N), 0.0, dtype=ct.float32)
    for k in range(0, ct.cdiv(K, TILE_K)):
        a_col_indices = k * TILE_K + ct.arange(TILE_K, dtype=ct.int32)
        a = ct.gather(A, (a_row_indices[:, None], a_col_indices[None, :]))

        b = ct.load(B, (expert_id, k, bid_n), shape=(1, TILE_K, TILE_N),
                    order=(0, 2, 1), padding_mode=zero_pad).reshape((TILE_K, TILE_N))

        accumulator = ct.mma(a, b, accumulator)

    if mul_routed_weight:
        moe_weight = ct.gather(topk_weights, token_ids)
        accumulator = accumulator * moe_weight[:, None]

    # Scatter result back into C
    c_col_indices = bid_n * TILE_N + ct.arange(TILE_N, dtype=ct.int32)
    accumulator = ct.astype(accumulator, C.dtype)
    ct.scatter(C, (token_ids[:, None], c_col_indices[None, :]), accumulator)


@ct.kernel
def silu_and_mul_kernel(A, B, C, TILE_N: ConstInt):
    """Element-wise kernel: computes SiLU(A) * B."""
    bid_m = ct.bid(0)
    ta = ct.load(A, (bid_m, 0), (1, TILE_N)).astype(ct.float32)
    tb = ct.load(B, (bid_m, 0), (1, TILE_N)).astype(ct.float32)

    # Sigmoid(ta)
    denom = ct.add(1, ct.exp(-ta), flush_to_zero=True)
    sigmoid_ta = ct.truediv(1.0, denom, flush_to_zero=True, rounding_mode=ct.RoundingMode.APPROX)

    # SiLU(ta) * tb
    silu_ta = ct.mul(ta, sigmoid_ta, flush_to_zero=True)
    tc = ct.mul(silu_ta, tb, flush_to_zero=True)

    ct.store(C, (bid_m, 0), tc.astype(C.dtype))


#=============================================================================
# Host-side helpers
#=============================================================================

def next_power_of_2(n):
    n -= 1
    n |= n >> 1
    n |= n >> 2
    n |= n >> 4
    n |= n >> 8
    n |= n >> 16
    n |= n >> 32
    n += 1
    return n


def moe_align_tile_size(topk_ids, tile_m, num_experts):
    """
    Sort, replicate, and pad token indices by expert so every expert processes
    a TILE_M-aligned number of tokens.

    Returns:
        sorted_token_ids: 1D array of token-replica indices sorted by expert,
            padded with sentinel value (num_tokens * topk).
        sorted_expert_ids: Expert id for each TILE_M block.
    """
    num_tokens, topk = topk_ids.shape
    total_tokens = num_tokens * topk

    # Do everything on CPU (small metadata arrays)
    flat_expert_ids = cp.asnumpy(topk_ids.reshape(-1))
    sorted_token_indices_cpu = np.argsort(flat_expert_ids, kind='stable').astype(np.int32)

    expert_token_counts_cpu = np.bincount(flat_expert_ids.astype(np.int32), minlength=num_experts)
    expert_block_counts_cpu = (expert_token_counts_cpu - 1 + tile_m) // tile_m
    total_blocks = int(expert_block_counts_cpu.sum())

    sorted_token_ids_cpu = np.full((total_blocks * tile_m,), total_tokens, dtype=np.int32)
    sorted_expert_ids_cpu = np.zeros(total_blocks, dtype=np.int32)

    current_block = 0
    current_token = 0
    for expert_id in range(num_experts):
        token_count = int(expert_token_counts_cpu[expert_id])
        block_count = int(expert_block_counts_cpu[expert_id])

        sorted_expert_ids_cpu[current_block:current_block + block_count] = expert_id
        sorted_token_start = current_block * tile_m
        sorted_token_ids_cpu[sorted_token_start:sorted_token_start + token_count] = \
            sorted_token_indices_cpu[current_token:current_token + token_count]

        current_token += token_count
        current_block += block_count

    return cp.asarray(sorted_token_ids_cpu), cp.asarray(sorted_expert_ids_cpu)


def invoke_fused_moe_kernel(A, B, C, topk_weights, sorted_token_ids, sorted_expert_ids,
                            mul_routed_weight, num_token_replicas, tile_m, tile_n, tile_k):
    m = sorted_token_ids.shape[0]
    n = B.shape[1]
    grid = (ceil(m / tile_m) * ceil(n / tile_n),)
    topk_weights_flat = topk_weights.reshape(-1)
    C_flat = C.reshape(-1, C.shape[-1])
    stream = cp.cuda.get_current_stream()
    ct.launch(stream, grid, fused_moe_kernel,
              (A, B, C_flat, topk_weights_flat, sorted_token_ids, sorted_expert_ids,
               num_token_replicas, mul_routed_weight, tile_m, tile_n, tile_k))


def invoke_silu_and_mul_kernel(AB, C):
    A_half, B_half = cp.split(AB, 2, axis=-1)
    # Make contiguous copies since split returns views
    A_half = cp.ascontiguousarray(A_half)
    B_half = cp.ascontiguousarray(B_half)
    stream = cp.cuda.get_current_stream()
    ct.launch(stream, (AB.shape[0],), silu_and_mul_kernel,
              (A_half, B_half, C, next_power_of_2(C.shape[-1])))


def cutile_moe(hidden_states, w1, w2, topk_weights, topk_ids, tile_m, tile_n, tile_k):
    """Run full MoE forward pass: expert matmul -> SiLU activation -> expert matmul."""
    num_tokens, hidden_size = hidden_states.shape
    num_experts, _, intermediate_size = w2.shape
    _, topk = topk_ids.shape

    intermediate_cache1 = cp.zeros((num_tokens, topk, intermediate_size * 2), dtype=hidden_states.dtype)
    intermediate_cache2 = cp.zeros((num_tokens * topk, intermediate_size), dtype=hidden_states.dtype)
    intermediate_cache3 = cp.zeros((num_tokens, topk, hidden_size), dtype=hidden_states.dtype)

    sorted_token_ids, sorted_expert_ids = moe_align_tile_size(topk_ids, tile_m, num_experts)

    # First matmul: hidden_states @ w1^T -> gate+up projection
    invoke_fused_moe_kernel(
        hidden_states, w1, intermediate_cache1, topk_weights,
        sorted_token_ids, sorted_expert_ids,
        mul_routed_weight=False, num_token_replicas=topk,
        tile_m=tile_m, tile_n=tile_n, tile_k=tile_k)

    # SiLU activation
    invoke_silu_and_mul_kernel(
        intermediate_cache1.reshape(-1, intermediate_cache1.shape[-1]),
        intermediate_cache2)

    # Second matmul: activated @ w2^T -> down projection (with routing weights)
    invoke_fused_moe_kernel(
        intermediate_cache2, w2, intermediate_cache3, topk_weights,
        sorted_token_ids, sorted_expert_ids,
        mul_routed_weight=True, num_token_replicas=1,
        tile_m=tile_m, tile_n=tile_n, tile_k=tile_k)

    return cp.sum(intermediate_cache3, axis=1)


#=============================================================================
# Reference implementation
#=============================================================================

def ref_moe(hidden_states, w1, w2, topk_weights, topk_ids):
    """Naive NumPy MoE for correctness checking."""
    hidden_states = cp.asnumpy(hidden_states).astype(np.float32)
    w1 = cp.asnumpy(w1).astype(np.float32)
    w2 = cp.asnumpy(w2).astype(np.float32)
    topk_weights = cp.asnumpy(topk_weights).astype(np.float32)
    topk_ids = cp.asnumpy(topk_ids)

    num_tokens = hidden_states.shape[0]
    num_experts = w1.shape[0]
    _, topk = topk_ids.shape

    final = np.zeros_like(hidden_states)

    for expert_id in range(num_experts):
        # Find which (token, k) pairs route to this expert
        token_indices, k_indices = np.where(topk_ids == expert_id)
        if len(token_indices) == 0:
            continue

        tokens = hidden_states[token_indices]  # (count, hidden_size)
        gate_up = w1[expert_id]  # (intermediate*2, hidden_size)
        gate_proj = gate_up[:gate_up.shape[0] // 2]
        up_proj = gate_up[gate_up.shape[0] // 2:]
        down_proj = w2[expert_id]  # (hidden_size, intermediate)

        gate_out = tokens @ gate_proj.T
        up_out = tokens @ up_proj.T
        # SiLU activation
        silu_out = gate_out / (1.0 + np.exp(-gate_out)) * up_out
        expert_out = silu_out @ down_proj.T

        weights = topk_weights[token_indices, k_indices]
        weighted = expert_out * weights[:, None]

        np.add.at(final, token_indices, weighted)

    return final


#=============================================================================
# Example harness
#=============================================================================

def prepare(*, benchmark=False,
            num_tokens=None, hidden_size=None, num_experts=None,
            intermediate_size=None, topk=None, dtype=np.float16):
    if num_tokens is None:
        num_tokens = 256 if benchmark else 48
    if hidden_size is None:
        hidden_size = 1024 if benchmark else 512
    if num_experts is None:
        num_experts = 32 if benchmark else 64
    if intermediate_size is None:
        intermediate_size = 2048 if benchmark else 1024
    if topk is None:
        topk = 8

    # Use uniform random to avoid float64 intermediates from randn
    hidden_states = (cp.random.random((num_tokens, hidden_size), dtype=np.float32).astype(dtype) - 0.5)
    w1 = (cp.random.random((num_experts, intermediate_size * 2, hidden_size), dtype=np.float32).astype(dtype) - 0.5) * 0.2
    w2 = (cp.random.random((num_experts, hidden_size, intermediate_size), dtype=np.float32).astype(dtype) - 0.5) * 0.2

    # Unique expert IDs per token (no repeats within a row)
    topk_ids_cpu = np.stack([
        np.random.permutation(num_experts)[:topk] for _ in range(num_tokens)
    ])
    topk_ids = cp.asarray(topk_ids_cpu.astype(np.int64))

    # Softmax routing weights
    topk_weights = cp.random.randn(num_tokens, topk).astype(np.float32)
    topk_weights = cp.exp(topk_weights)
    topk_weights = topk_weights / topk_weights.sum(axis=1, keepdims=True)
    topk_weights = topk_weights.astype(dtype)

    return {
        "hidden_states": hidden_states,
        "w1": w1,
        "w2": w2,
        "topk_weights": topk_weights,
        "topk_ids": topk_ids,
        "num_tokens": num_tokens,
        "hidden_size": hidden_size,
        "num_experts": num_experts,
        "intermediate_size": intermediate_size,
        "topk": topk,
        "tile_m": 128,
        "tile_n": 128,
        "tile_k": 64,
    }


def run(data, *, nruns=1, warmup=0):
    hs = data["hidden_states"]
    w1, w2 = data["w1"], data["w2"]
    tw, ti = data["topk_weights"], data["topk_ids"]
    tm, tn, tk = data["tile_m"], data["tile_n"], data["tile_k"]

    stream = cp.cuda.get_current_stream()

    for _ in range(warmup):
        cutile_moe(hs, w1, w2, tw, ti, tm, tn, tk)
    cp.cuda.runtime.deviceSynchronize()

    times = []
    for _ in range(nruns):
        start = cp.cuda.Event()
        end = cp.cuda.Event()
        start.record(stream)
        out = cutile_moe(hs, w1, w2, tw, ti, tm, tn, tk)
        end.record(stream)
        end.synchronize()
        times.append(cp.cuda.get_elapsed_time(start, end))

    return {"out": out, "times": times}


def verify(data, result):
    expected = ref_moe(data["hidden_states"], data["w1"], data["w2"],
                       data["topk_weights"], data["topk_ids"])
    actual = cp.asnumpy(result["out"]).astype(np.float32)
    assert np.allclose(actual, expected, rtol=1e-1, atol=1e-1), \
        f"MoE incorrect! max diff: {np.max(np.abs(actual - expected))}"


def metric(data):
    # Two matmuls per token per expert-selection:
    # matmul1: (num_tokens*topk, hidden) @ (hidden, intermediate*2) = 2*tokens*topk*hidden*intermediate*2
    # matmul2: (num_tokens*topk, intermediate) @ (intermediate, hidden) = 2*tokens*topk*intermediate*hidden
    # Total: 2 * num_tokens * topk * hidden * intermediate * 3  (gate+up = 2x, down = 1x)
    nt = data["num_tokens"]
    topk = data["topk"]
    hs = data["hidden_size"]
    inter = data["intermediate_size"]
    flops = 2 * nt * topk * hs * inter * 3
    return flops, "TFLOPS"


#=============================================================================
# Main
#=============================================================================

def test_moe(num_tokens, hidden_size, num_experts, intermediate_size, topk,
             tile_m=128, tile_n=128, tile_k=64, dtype=np.float16, name=None):
    name = name or (f"moe tokens={num_tokens}, hidden={hidden_size}, experts={num_experts}, "
                    f"inter={intermediate_size}, topk={topk}, {dtype.__name__}")
    print(f"--- {name} ---")
    data = prepare(num_tokens=num_tokens, hidden_size=hidden_size, num_experts=num_experts,
                   intermediate_size=intermediate_size, topk=topk, dtype=dtype)
    data["tile_m"] = tile_m
    data["tile_n"] = tile_n
    data["tile_k"] = tile_k
    result = run(data)
    verify(data, result)
    print("  passed")


def main():
    print("--- cuTile Mixture of Experts Examples ---\n")

    test_moe(48, 512, 64, 1024, 8)
    test_moe(128, 512, 32, 1024, 4)
    test_moe(64, 1024, 64, 2048, 8)

    print("\n--- All MoE examples completed ---")


if __name__ == "__main__":
    main()

File: examples/moe.jl
# Mixture of Experts example - Julia port of cuTile Python's MoE.py sample
#
# Data layout: Julia column-major shapes are reversed from Python's row-major so
# that both produce identical Tile IR. The fast-varying dimension is listed first:
#   Python A(num_tokens, K)    → Julia A(K, num_tokens)
#   Python B(experts, N, K)    → Julia B(K, N, experts)
#   Python C(total_tokens, N)  → Julia C(N, total_tokens)
#
# SPDX-License-Identifier: Apache-2.0

using CUDA
using Random: randperm
import cuTile as ct

#=============================================================================
 Helper: 2D swizzle (same pattern as matmul.jl)
=============================================================================#

@inline function swizzle_2d(M, N, tm, tn, GROUP_SIZE_M, bid)
    num_bid_m = cld(M, Int32(tm))
    num_bid_n = cld(N, Int32(tn))
    num_bid_in_group = Int32(GROUP_SIZE_M) * num_bid_n
    group_id = fld(bid, num_bid_in_group)
    first_bid_m = group_id * Int32(GROUP_SIZE_M)
    group_size_m = min(num_bid_m - first_bid_m, Int32(GROUP_SIZE_M))
    bid_m = first_bid_m + rem(bid, group_size_m)
    bid_n = fld(rem(bid, num_bid_in_group), group_size_m)
    return bid_m, bid_n
end

#=============================================================================
 Kernels
=============================================================================#

# Fused MoE kernel: multiplies routed tokens by their assigned expert weights.
#
# A: (K, num_tokens) - input tokens/activations (K contiguous)
# B: (K, N, num_experts) - expert weight matrices (K contiguous)
# C: (N, total_tokens) - output (N contiguous)
# topk_weights: (total_tokens,) - flattened routing weights
# sorted_token_ids: (M_padded,) - 1-indexed token replica IDs, sorted by expert
# sorted_expert_ids: (num_blocks,) - 1-indexed expert ID per TILE_M block
function fused_moe_kernel(A::ct.TileArray{T, 2}, B::ct.TileArray{T, 3},
                          C::ct.TileArray{T, 2},
                          topk_weights::ct.TileArray{T, 1},
                          sorted_token_ids::ct.TileArray{Int32, 1},
                          sorted_expert_ids::ct.TileArray{Int32, 1},
                          num_token_replicas::Int, mul_routed_weight::Bool,
                          TILE_M::Int, TILE_N::Int, TILE_K::Int) where {T}
    M = size(sorted_token_ids, 1)
    K = size(B, 1)
    N = size(B, 2)

    bid = ct.bid(1) - Int32(1)  # 0-indexed for swizzle
    bid_m, bid_n = swizzle_2d(M, N, TILE_M, TILE_N, Int32(8), bid)

    # Gather 1-indexed token IDs for this block
    token_id_indices = bid_m * Int32(TILE_M) .+ ct.arange(TILE_M)
    token_ids = ct.gather(sorted_token_ids, token_id_indices)

    # Map 1-indexed flat token_id to 1-indexed column in A
    # token_id k → original token = (k-1) ÷ num_token_replicas + 1
    a_tok_indices = (token_ids .- Int32(1)) .÷ Int32(num_token_replicas) .+ Int32(1)

    # Expert for this block (scalar, 1-indexed tile index for load)
    expert_id = sorted_expert_ids[bid_m + Int32(1)]

    acc = zeros(Float32, TILE_N, TILE_M)
    num_k = cld(K, Int32(TILE_K))

    for k in Int32(1):num_k
        # 1-indexed row indices into A's K dimension
        a_k_indices = (k - Int32(1)) * Int32(TILE_K) .+ ct.arange(TILE_K)

        # A is (K, num_tokens): dim 1 = K, dim 2 = tokens
        a = ct.gather(A, (reshape(a_k_indices, (TILE_K, 1)),
                          reshape(a_tok_indices, (1, TILE_M))))  # (TILE_K, TILE_M)

        # B is (K, N, num_experts): load (TILE_N, TILE_K) slice for this expert
        # order=(2,1,3) folds the transpose into the load via dim_map, matching
        # Python cuTile's order=(0,2,1) and avoiding an explicit permute in Tile IR.
        b = ct.load(B; index=(bid_n + Int32(1), k, expert_id),
                    shape=(TILE_N, TILE_K, 1), order=(2, 1, 3),
                    padding_mode=ct.PaddingMode.Zero)
        b = reshape(b, (TILE_N, TILE_K))

        # acc(N,M) += b(N,K) @ a(K,M)
        acc = muladd(b, a, acc)
    end

    if mul_routed_weight
        moe_weight = convert(ct.Tile{Float32}, ct.gather(topk_weights, token_ids))
        acc = acc .* reshape(moe_weight, (1, TILE_M))
    end

    # Scatter result to C at token_id positions
    # C is (N, total_tokens): dim 1 = N, dim 2 = tokens
    c_n_indices = bid_n * Int32(TILE_N) .+ ct.arange(TILE_N)  # 1-indexed
    output = convert(ct.Tile{T}, acc)
    ct.scatter(C, (reshape(c_n_indices, (TILE_N, 1)),
                   reshape(token_ids, (1, TILE_M))), output)

    return nothing
end


# Element-wise SiLU activation: computes SiLU(A) * B
# Arrays are (dim, total_tokens): each block processes one column (= one token).
function silu_and_mul_kernel(A::ct.TileArray{T, 2}, B::ct.TileArray{T, 2},
                             C::ct.TileArray{T, 2}, TILE_N::Int) where {T}
    bid_m = ct.bid(1)
    ta = convert(ct.Tile{Float32}, ct.load(A; index=(Int32(1), bid_m), shape=(TILE_N, 1)))
    tb = convert(ct.Tile{Float32}, ct.load(B; index=(Int32(1), bid_m), shape=(TILE_N, 1)))

    # SiLU(x) = x * sigmoid(x) = x / (1 + exp(-x))
    denom = 1.0f0 .+ exp.(.-ta)
    sigmoid_ta = 1.0f0 ./ denom
    silu_ta = ta .* sigmoid_ta
    tc = silu_ta .* tb

    ct.store(C; index=(Int32(1), bid_m), tile=convert(ct.Tile{T}, tc))
    return nothing
end


#=============================================================================
 Host-side helpers
=============================================================================#

"""
Sort, replicate, and pad token indices by expert so every expert processes
a TILE_M-aligned number of tokens.

`topk_ids` is (num_tokens, topk) with 1-indexed expert IDs.
Returns GPU arrays with 1-indexed values.
"""
function moe_align_tile_size(topk_ids::Matrix{Int}, tile_m::Int, num_experts::Int)
    num_tokens, topk = size(topk_ids)
    total_tokens = num_tokens * topk

    # Flatten in Python-style order (slot varies fastest within each token).
    # permutedims gives (topk, num_tokens); vec in column-major = slot-first ordering.
    flat_expert_ids = vec(permutedims(topk_ids))

    sorted_perm = sortperm(flat_expert_ids; stable=true)  # 1-indexed

    expert_token_counts = zeros(Int, num_experts)
    for eid in flat_expert_ids
        expert_token_counts[eid] += 1
    end
    expert_block_counts = cld.(expert_token_counts, tile_m)
    total_blocks = sum(expert_block_counts)

    # Sentinel value: total_tokens + 1 (out-of-bounds for 1-indexed arrays)
    sorted_token_ids = fill(Int32(total_tokens + 1), total_blocks * tile_m)
    sorted_expert_ids = zeros(Int32, total_blocks)

    current_block = 0
    current_token = 0
    for eid in 1:num_experts
        tc = expert_token_counts[eid]
        bc = expert_block_counts[eid]

        for i in 1:bc
            sorted_expert_ids[current_block + i] = Int32(eid)
        end

        start = current_block * tile_m + 1
        for i in 1:tc
            sorted_token_ids[start + i - 1] = Int32(sorted_perm[current_token + i])
        end

        current_token += tc
        current_block += bc
    end

    return CuArray(sorted_token_ids), CuArray(sorted_expert_ids)
end


function invoke_fused_moe_kernel(A, B, C, topk_weights, sorted_token_ids, sorted_expert_ids;
                                 mul_routed_weight, num_token_replicas, tile_m, tile_n, tile_k)
    m = length(sorted_token_ids)
    n = size(B, 2)
    grid = cld(m, tile_m) * cld(n, tile_n)

    # Flatten in Python-style order (slot varies fastest) to match token_id ordering
    topk_weights_flat = vec(permutedims(topk_weights))

    # C is 3D (dim, topk, num_tokens) → flatten last two dims to (dim, total_tokens)
    C_flat = reshape(C, size(C, 1), size(C, 2) * size(C, 3))

    ct.launch(fused_moe_kernel, grid,
              A, B, C_flat, topk_weights_flat,
              sorted_token_ids, sorted_expert_ids,
              num_token_replicas, ct.Constant(mul_routed_weight),
              ct.Constant(tile_m), ct.Constant(tile_n), ct.Constant(tile_k))
end


function invoke_silu_and_mul_kernel(AB, C)
    inter = size(C, 1)  # C is (intermediate, total_tokens)
    total = size(AB, 2)

    # Split AB(inter*2, total) into gate and up halves along dim 1.
    #A_half = AB[1:inter, :]
    #B_half = AB[inter+1:2*inter, :]
    # FIXME: CUDA.jl's CuArray indexing (AB[1:inter, :]) uses a slow generic kernel.
    # Use unsafe_copy2d! (cuMemcpy2D) for hardware-accelerated pitched 2D copy instead.
    T = eltype(AB)
    A_half = similar(AB, inter, total)
    B_half = similar(AB, inter, total)
    src_pitch = size(AB, 1) * sizeof(T)
    dst_pitch = inter * sizeof(T)
    CUDA.unsafe_copy2d!(pointer(A_half), CUDA.DeviceMemory,
                        pointer(AB), CUDA.DeviceMemory,
                        inter, total; srcPitch=src_pitch, dstPitch=dst_pitch,
                        async=true)
    CUDA.unsafe_copy2d!(pointer(B_half), CUDA.DeviceMemory,
                        pointer(AB) + inter * sizeof(T), CUDA.DeviceMemory,
                        inter, total; srcPitch=src_pitch, dstPitch=dst_pitch,
                        async=true)

    tile_n = nextpow(2, inter)
    ct.launch(silu_and_mul_kernel, total,
              A_half, B_half, C, ct.Constant(tile_n))
end


function cutile_moe(hidden_states::CuArray{T}, w1, w2, topk_weights, topk_ids,
                    tile_m, tile_n, tile_k) where {T}
    hidden_size, num_tokens = size(hidden_states)  # (hidden, num_tokens)
    intermediate_size = size(w2, 1)                # w2 is (inter, hidden, experts)
    num_experts = size(w1, 3)
    _, topk = size(topk_ids)
    total_tokens = num_tokens * topk

    # Intermediate caches: reversed from Python for column-major
    # Python (num_tokens, topk, dim) → Julia (dim, topk, num_tokens)
    cache1 = CUDA.zeros(T, intermediate_size * 2, topk, num_tokens)
    cache2 = CUDA.zeros(T, intermediate_size, total_tokens)
    cache3 = CUDA.zeros(T, hidden_size, topk, num_tokens)

    sorted_token_ids, sorted_expert_ids = moe_align_tile_size(
        Array(topk_ids), tile_m, num_experts)

    # First matmul: hidden_states @ w1^T → gate+up projection
    invoke_fused_moe_kernel(
        hidden_states, w1, cache1,
        topk_weights, sorted_token_ids, sorted_expert_ids;
        mul_routed_weight=false, num_token_replicas=topk,
        tile_m, tile_n, tile_k)

    # SiLU activation: SiLU(gate) * up
    # cache1 is (inter*2, topk, num_tokens) → flatten to (inter*2, total_tokens)
    cache1_flat = reshape(cache1, intermediate_size * 2, total_tokens)
    invoke_silu_and_mul_kernel(cache1_flat, cache2)

    # Second matmul: activated @ w2^T → down projection, with routing weights
    invoke_fused_moe_kernel(
        cache2, w2, cache3,
        topk_weights, sorted_token_ids, sorted_expert_ids;
        mul_routed_weight=true, num_token_replicas=1,
        tile_m, tile_n, tile_k)

    # Sum over topk slots (second dimension) for each token
    # cache3 is (hidden, topk, num_tokens) → sum over topk
    return dropdims(sum(cache3; dims=2); dims=2)
end


#=============================================================================
 Reference implementation
=============================================================================#

function ref_moe(hidden_states, w1, w2, topk_weights, topk_ids)
    hs = Float32.(Array(hidden_states))     # (hidden_size, num_tokens)
    w1_cpu = Float32.(Array(w1))            # (hidden_size, intermediate_size*2, num_experts)
    w2_cpu = Float32.(Array(w2))            # (intermediate_size, hidden_size, num_experts)
    tw_cpu = Float32.(Array(topk_weights))  # (num_tokens, topk)
    ti_cpu = Array(topk_ids)                # (num_tokens, topk)

    hidden_size, num_tokens = size(hs)
    intermediate_size = size(w2_cpu, 1)
    num_experts = size(w1_cpu, 3)

    result = zeros(Float32, hidden_size, num_tokens)

    for eid in 1:num_experts
        positions = findall(==(eid), ti_cpu)
        isempty(positions) && continue

        token_indices = [p[1] for p in positions]
        slot_indices = [p[2] for p in positions]

        # hs is (hidden, num_tokens) → tokens are columns
        tokens = hs[:, token_indices]  # (hidden, count)

        w1e = w1_cpu[:, :, eid]  # (hidden, intermediate*2)
        gate_proj = w1e[:, 1:intermediate_size]           # (hidden, intermediate)
        up_proj = w1e[:, intermediate_size+1:end]         # (hidden, intermediate)

        gate_out = gate_proj' * tokens   # (intermediate, count)
        up_out = up_proj' * tokens       # (intermediate, count)
        silu_out = gate_out ./ (1.0f0 .+ exp.(.-gate_out)) .* up_out

        down_proj = w2_cpu[:, :, eid]   # (intermediate, hidden)
        expert_out = down_proj' * silu_out  # (hidden, count)

        for (i, (tok, slot)) in enumerate(zip(token_indices, slot_indices))
            result[:, tok] .+= expert_out[:, i] .* tw_cpu[tok, slot]
        end
    end

    return result
end


#=============================================================================
 Example harness
=============================================================================#

function prepare(; benchmark::Bool=false,
                  num_tokens::Int = benchmark ? 256 : 48,
                  hidden_size::Int = benchmark ? 1024 : 512,
                  num_experts::Int = benchmark ? 32 : 64,
                  intermediate_size::Int = benchmark ? 2048 : 1024,
                  topk::Int = 8,
                  T::DataType = Float16)
    # hidden_states is (hidden, num_tokens): hidden contiguous
    hidden_states = CUDA.rand(T, hidden_size, num_tokens) .- T(0.5)
    w1 = (CUDA.rand(T, hidden_size, intermediate_size * 2, num_experts) .- T(0.5)) .* T(0.2)
    w2 = (CUDA.rand(T, intermediate_size, hidden_size, num_experts) .- T(0.5)) .* T(0.2)

    # Unique expert IDs per token (1-indexed)
    topk_ids = Matrix{Int}(undef, num_tokens, topk)
    for i in 1:num_tokens
        topk_ids[i, :] = randperm(num_experts)[1:topk]
    end

    # Softmax routing weights
    raw = rand(Float32, num_tokens, topk)
    raw_exp = exp.(raw)
    topk_weights = T.(raw_exp ./ sum(raw_exp; dims=2))

    return (;
        hidden_states,
        w1, w2,
        topk_weights = CuArray(topk_weights),
        topk_ids = CuArray(topk_ids),
        num_tokens, hidden_size, num_experts, intermediate_size, topk,
        tile_m = 128, tile_n = 128, tile_k = 64
    )
end


function run(data; nruns::Int=1, warmup::Int=0)
    (; hidden_states, w1, w2, topk_weights, topk_ids, tile_m, tile_n, tile_k) = data

    CUDA.@sync for _ in 1:warmup
        cutile_moe(hidden_states, w1, w2, topk_weights, topk_ids, tile_m, tile_n, tile_k)
    end

    times = Float64[]
    out = nothing
    for _ in 1:nruns
        t = CUDA.@elapsed begin
            out = cutile_moe(hidden_states, w1, w2, topk_weights, topk_ids,
                             tile_m, tile_n, tile_k)
        end
        push!(times, t * 1000)  # ms
    end

    return (; out, times)
end


function verify(data, result)
    expected = ref_moe(data.hidden_states, data.w1, data.w2, data.topk_weights, data.topk_ids)
    actual = Float32.(Array(result.out))
    @assert isapprox(actual, expected; rtol=1e-2, atol=1e-2) "MoE incorrect! max diff: $(maximum(abs.(actual .- expected)))"
end


function metric(data)
    nt = data.num_tokens
    topk = data.topk
    hs = data.hidden_size
    inter = data.intermediate_size
    return 2 * nt * topk * hs * inter * 3, "TFLOPS"
end


#=============================================================================
 Main
=============================================================================#

function test_moe(num_tokens, hidden_size, num_experts, intermediate_size, topk;
                  tile_m=128, tile_n=128, tile_k=64, T=Float16, name=nothing)
    name = something(name, "moe tokens=$num_tokens, hidden=$hidden_size, experts=$num_experts, " *
                           "inter=$intermediate_size, topk=$topk, $T")
    println("--- $name ---")
    data = prepare(; num_tokens, hidden_size, num_experts, intermediate_size, topk, T)
    result = run(data)
    verify(data, result)
    println("  passed")
end


function main()
    println("--- cuTile Mixture of Experts Examples ---\n")

    test_moe(48, 512, 64, 1024, 8)
    test_moe(128, 512, 32, 1024, 4)
    test_moe(64, 1024, 64, 2048, 8)

    println("\n--- All MoE examples completed ---")
end

isinteractive() || main()
